Discover Plex clients using GDM

jjlawren opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds HA-initiated GDM network scanning as another method to discover controllable Plex clients. GDM uses UDP broadcasts to discover other Plex servers/clients. As some Plex servers are in a different subnet (read: Docker), this client discovery is sometimes unreliable. Throttles the updates to avoid unnecessary scans.

Current order for Plex client discovery:
1. Provided by Plex server (does its own GDM discovery)
2. GDM broadcast from HA
3. plex.tv
4. Active playback session from Plex server (usually not controllable)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jjlawren committed:
Add local network scanning for Plex clients

jjlawren committed:
Update tests for GDM client discovery

jjlawren commented:
Moving this to a draft for now. After some thought it would be better to set up a single GDM scanner for all configured Plex servers instead of one for each. I'll need to think about how this would affect client scanning as the GDM payload doesn't indicate which server(s) a discovered client has permissions to access.

jjlawren committed:
Make single shared GDM scanner for all config entries

jjlawren committed:
flake & black

jjlawren commented:
The use case that I haven't conclusively tested for is when an HA instance can reach local Plex clients that aren't associated with any of the Plex servers which HA has access to. Those connections _should_ fail gracefully. I'm also not convinced that's a common enough scenario to worry about for now.

balloob commented on homeassistant/components/plex/server.py:
```diff
@@ -339,6 +351,22 @@ def connect_to_resource(resource):
+                baseurl = f"http://{gdm_client['from'][0]}:{gdm_client['data']['Port']}"
+                name = gdm_client["data"]["Name"]
+                try:
+                    device = await self.hass.async_add_executor_job(
```
We're doing an `await` inside a loop, which means we are not going to attempt to connect to the next one until the current one is processed. Is that necessary?

jjlawren replied:
I think this should happen infrequently enough and quickly enough that it shouldn't matter.

MartinHjelmare replied:
A better solution is to extract the iteration and calls to a function and schedule that whole function on the executor.

MartinHjelmare commented on tests/components/plex/test_config_flow.py:
```diff
@@ -434,10 +442,14 @@ async def test_option_flow_new_users_available(
+    mock_plex_server = await setup_plex_server(config_entry=entry, disable_gdm=False)
+
+    # Multiple debouncers with cooldowns involved, easier to trigger manually
+    async_dispatcher_send(hass, PLEX_GDM_CLIENT_SCAN_SIGNAL)
```
As soon as we're relying on integration details in the tests they become less robust. It would be better to call the plex websocket callback to trigger this signal. We seem to do that already below. Why isn't that enough?

MartinHjelmare replied:
We can move time forward as needed by firing a time changed event.
https://github.com/home-assistant/core/blob/5dbb5f12eba9f88a507bf6ca480356e2fc717c30/tests/common.py#L287-L288

jjlawren replied:
Seems this was a race between dispatching a call to begin the GDM scan and checking the result of that scan before it was scheduled/completed.

jjlawren committed:
Refactor GDM updates & client connections

Run GDM scan before each update via debouncer
Connect to all new clients from a single executor job
Cache connected clients to avoid unnecessary network calls
Don't attempt to reconnect to unauthorized clients
Use a temporary token to control clients
Remove integration-specific test code

MartinHjelmare reviewed (approved):
Looks good!

jjlawren commented on homeassistant/components/plex/server.py:
```diff
@@ -292,9 +291,10 @@ async def _async_update_platforms(self):
-        async_dispatcher_send(self.hass, PLEX_GDM_CLIENT_SCAN_SIGNAL)
-
         def process_device(source, device):
+            if device is None:
```
After looking at this again I don't think this check inside the function is necessary. It's a leftover from a different attempt during the refactor.

jjlawren replied:
Addressed in #41844.
